Wrap indirect indexing on CUDA

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105055

Lifting this to CPU should be rather easy. @jgong5
Partially fixes https://github.com/pytorch/pytorch/issues/97365. I'd wait to close that issue once this works on CPU as well.

This fix works with dynamic shapes as well.


@voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

lezcano committed:
Wrap indirect indexing on CUDA

Lifting this to CPU should be rather easy. cc @jgong5

Partially fixes https://github.com/pytorch/pytorch/issues/97365

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1201,6 +1203,26 @@ def __call__(self):
+            )
+
+            new_var.update_on_args("index_wrap", (var,), {})
+            var = new_var
```
Should `index_wrap` be it's own operator? `_unsafe_index` for example probably shouldn't generate wrapping.

peterbell10 replied:
Either a separate operator, or a `wrap=True` argument that we can selectively disable.

lezcano replied:
what about freeride off the `check` argument, and just generate this `check=True`. At the moment we're using `check=False` to circumvent limitations of the bound variable analysis. In all those cases, the indices are always positive so...

peterbell10 replied:
I can think of examples like backwards functions where you might still need wrapping, but could prove that the indices are in-bounds because they've already been checked elsewhere in the program.

lezcano replied:
Fair enough. Let's write that optimisation in another PR though. I can add the `if check` in this one if you want, though, as it'd be correct with the uses we currently have of `_unsafe_index` and `index_put`.

lezcano committed:
Update on "Wrap indirect indexing on CUDA"


Lifting this to CPU should be rather easy. cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 
Partially fixes https://github.com/pytorch/pytorch/issues/97365. I'd wait to close that issue once this works on CPU as well.


voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8



[ghstack-poisoned]

lezcano committed:
Update on "Wrap indirect indexing on CUDA"


Lifting this to CPU should be rather easy. cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 
Partially fixes https://github.com/pytorch/pytorch/issues/97365. I'd wait to close that issue once this works on CPU as well.


voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8



[ghstack-poisoned]

lezcano committed:
Update on "Wrap indirect indexing on CUDA"


Lifting this to CPU should be rather easy. jgong5
Partially fixes https://github.com/pytorch/pytorch/issues/97365. I'd wait to close that issue once this works on CPU as well.


voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8



[ghstack-poisoned]

peterbell10 commented on test/inductor/test_torchinductor.py:
```diff
@@ -6772,6 +6772,49 @@ def fn(x: torch.Tensor) -> torch.Tensor:
+                return A[b]
+
+            a = torch.rand(1024, device="cuda")
+            b = -torch.arange(start=-a.numel() + 1, end=-1, device="cuda")
```
These indices are positive, they don't test the wrapping.

peterbell10 commented on test/inductor/test_torchinductor.py:
```diff
@@ -6772,6 +6772,49 @@ def fn(x: torch.Tensor) -> torch.Tensor:
+
+            def flip_with_index(A):
+                b = -torch.arange(start=-a.numel() + 1, end=-1, device="cuda")
+                return A[b]
```
We optimize away the `indirect_indexing` call, so this never gets wrapped.

peterbell10 replied:
@lezcano you don't seem to have fixed this, you just removed the test case. To fix it I think you need to either do range analysis in the IndexPropagation pass, or add wrapping to the sympy expression,

lezcano replied:
ah, right, I misunderstood you. Now I see what you mean, I'll fix that.

Skylion007 commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -83,9 +83,12 @@ def __contains__(self, x):
+        return self & other
+
+    # Intersection
+    def __and__(self, other):
```
Return type for these functions should always be "ValueRange", right?

lezcano replied:
yep, similar to `dict`.

Skylion007 replied:
Might want to explicitly typehint that. :)

lezcano committed:
Update on "Wrap indirect indexing on CUDA"


Lifting this to CPU should be rather easy. jgong5
Partially fixes https://github.com/pytorch/pytorch/issues/97365. I'd wait to close that issue once this works on CPU as well.


voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8



[ghstack-poisoned]

lezcano commented:
@peterbell10 rehasehed the tests, and now we test that we are actually generating warping when we claim we do.

jgong5 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1202,6 +1204,29 @@ def __call__(self):
+            if var.bounds != ValueRanges.unknown():
+                # Take the negative part of the bound and add size to it
+                # Then take union of that and the positive part
+                # This is a tighter bound than that of a generic ops.where, as we have info on the conde
```
nit: conde -> code ;-)

lezcano replied:
cond* (as in the condition from the `where`) but yep :D

jgong5 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1202,6 +1204,29 @@ def __call__(self):
+                    pos = var.bounds & ValueRanges(0, sympy.oo)
+                    new_bounds = new_bounds | pos
+
+            new_var = self.cse.generate(
+                self.compute,
+                f"tl.where({var} < 0, {var} + {str_size}, {var})",
+                bounds=new_bounds,
+            )
```
As you commented, better to move this to `common.py` so that it can be applied to both triton and cpp backends? I guess if this is changed to the `ops` calls, it becomes general? We can submit a follow-up PR for that if you prefer.

lezcano replied:
I just did it this way, because that way I could call `update_on_args` in order to get the masks updated appropriately (note that the logic for when to assign the masks to `var` on `update_on_args` is less than good). If that is taken care of, then yes, calling `ops.` here would allow for this code to be generic.

jgong5 reviewed (commented):
> Lifting this to CPU should be rather easy. @jgong5
Partially fixes https://github.com/pytorch/pytorch/issues/97365. I'd wait to close that issue once this works on CPU as well.

@blzheng is working on fixing a similar issue https://github.com/pytorch/pytorch/issues/102064 with PR https://github.com/pytorch/pytorch/pull/102602. But I guess the fix in this PR can address that problem as well and is more complete. Probably @blzheng can provide a follow-up PR to cover CPP backend.

lezcano commented:
Yep, I just started this PR because it was reasonably straightforward and https://github.com/pytorch/pytorch/pull/102602 hadn't been updated in the last month.

lezcano committed:
Update on "Wrap indirect indexing on CUDA"


Lifting this to CPU should be rather easy. jgong5
Partially fixes https://github.com/pytorch/pytorch/issues/97365. I'd wait to close that issue once this works on CPU as well.


voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8



[ghstack-poisoned]

lezcano committed:
Update on "Wrap indirect indexing on CUDA"


Lifting this to CPU should be rather easy. jgong5
Partially fixes https://github.com/pytorch/pytorch/issues/97365. I'd wait to close that issue once this works on CPU as well.


voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8



[ghstack-poisoned]

lezcano committed:
Update on "Wrap indirect indexing on CUDA"


Lifting this to CPU should be rather easy. jgong5
Partially fixes https://github.com/pytorch/pytorch/issues/97365. I'd wait to close that issue once this works on CPU as well.


voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8



[ghstack-poisoned]

lezcano committed:
Update on "Wrap indirect indexing on CUDA"


Lifting this to CPU should be rather easy. jgong5
Partially fixes https://github.com/pytorch/pytorch/issues/97365. I'd wait to close that issue once this works on CPU as well.


voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8



[ghstack-poisoned]

lezcano committed:
Update on "Wrap indirect indexing on CUDA"


Lifting this to CPU should be rather easy. jgong5
Partially fixes https://github.com/pytorch/pytorch/issues/97365. I'd wait to close that issue once this works on CPU as well.


voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8



[ghstack-poisoned]

lezcano committed:
Update on "Wrap indirect indexing on CUDA"


Lifting this to CPU should be rather easy. jgong5
Partially fixes https://github.com/pytorch/pytorch/issues/97365. I'd wait to close that issue once this works on CPU as well.


voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8



[ghstack-poisoned]

peterbell10 commented on torch/_inductor/ir.py:
```diff
@@ -1849,6 +1849,7 @@ def indexer(index):
             for idx, stride, sz in zip(index, self.stride, self.size):
                 if sz != 1:
+                    idx = idx + Where(idx >= 0, 0, sz)
                     result = result + idx * stride
```
Don't think this is the right place for this. Probably this belongs in `index_propagation.py` when we optimize out `ops.index_expr(var, size)`, that way only user-generated indexing should get wrapped.

Also should this not be `sympy.Ge(idx, 0)`?

lezcano replied:
Yes, sorry, should have moved this to WIP. I just wanted to push these chnages to run CI, but I was aware that this was not correct.

lezcano committed:
Update on "Wrap indirect indexing on CUDA"


Lifting this to CPU should be rather easy. jgong5
Partially fixes https://github.com/pytorch/pytorch/issues/97365. I'd wait to close that issue once this works on CPU as well.


voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8



[ghstack-poisoned]

lezcano commented on torch/_inductor/codegen/common.py:
```diff
@@ -310,6 +310,9 @@ def _print_Mod(self, expr):
         return self._print_FloorDiv(expr)  # type: ignore[attr-defined]
 
+    def _print_GreaterThan(self, expr):
+        return " >= ".join(map(self.paren, map(self._print, expr.args)))
```
I hadn't even realised lol. I blindly copied the name of the class after checkign that `Ge` didn't work.

lezcano committed:
Update on "Wrap indirect indexing on CUDA"


Lifting this to CPU should be rather easy. jgong5
Partially fixes https://github.com/pytorch/pytorch/issues/97365. I'd wait to close that issue once this works on CPU as well.

This fix works with dynamic shapes as well.


voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8



[ghstack-poisoned]

lezcano commented:
@pytorchbot merge

lezcano committed:
Update on "Wrap indirect indexing on CUDA"


Lifting this to CPU should be rather easy. jgong5
Partially fixes https://github.com/pytorch/pytorch/issues/97365. I'd wait to close that issue once this works on CPU as well.

This fix works with dynamic shapes as well.


voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

Skylion007 commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -82,9 +82,12 @@ def __contains__(self, x):
+        return self & other
+
+    # Intersection
+    def __and__(self, other):
```
```suggestion
    def __and__(self, other) -> "ValueRange":
```

Skylion007 commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -96,9 +99,16 @@ def tighten(self, other: "ValueRanges"):
-    def __and__(self, other):
-        return ValueRanges(lower=max(self.lower, other.lower), upper=min(self.upper, other.upper))
+    # Union
+    def __or__(self, other):
```
```suggestion
    def __or__(self, other) -> "ValueRange":
```

lezcano committed:
Update on "Wrap indirect indexing on CUDA"


Lifting this to CPU should be rather easy. jgong5
Partially fixes https://github.com/pytorch/pytorch/issues/97365. I'd wait to close that issue once this works on CPU as well.

This fix works with dynamic shapes as well.


voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

lezcano committed:
Update on "Wrap indirect indexing on CUDA"


Lifting this to CPU should be rather easy. jgong5
Partially fixes https://github.com/pytorch/pytorch/issues/97365. I'd wait to close that issue once this works on CPU as well.

This fix works with dynamic shapes as well.


voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

lezcano committed:
Update on "Wrap indirect indexing on CUDA"


Lifting this to CPU should be rather easy. jgong5
Partially fixes https://github.com/pytorch/pytorch/issues/97365. I'd wait to close that issue once this works on CPU as well.

This fix works with dynamic shapes as well.


voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot revert -m "Causes failure in inductor_torchbench" -c ghfirst

peterbell10 commented:
@lezcano looks like the CI failure here is real.

lezcano committed:
Update on "Wrap indirect indexing on CUDA"


Lifting this to CPU should be rather easy. jgong5
Partially fixes https://github.com/pytorch/pytorch/issues/97365. I'd wait to close that issue once this works on CPU as well.

This fix works with dynamic shapes as well.


voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

lezcano commented:
Ok, moved from failed to run to not accurate. I will investigate the not accurate part as well.

lezcano commented:
@pytorchbot merge

lezcano commented:
@pytorchbot merge -i

lezcano commented:
@pytorchbot drci

lezcano commented:
@pytorchbot merge -i
